### PR TITLE
Fix #988 - Support mutual SSL in API level 

### LINKED ISSUE #988: Support mutual SSL in API level
### Describe your problem(s)

Support Mutual SSL at API level. 

### Describe your solution

Configure Mutual SSL verify client value to optional/required in API level.
Use x-wso2-transport-security extension.
x-wso2-transport-security:
  mutualssl: "mandatory" # support mandatory and optional options here

### How will you implement it

---

### LINKED ISSUE #1269: Per API mutual ssl cert validation
### Describe your problem(s)
In microgateway we can specify to enable the mutual ssl per API. But we can not specify the set of certificates that allowed to access the API. The certificate against the API validation is not done.
At the moment having a certificate trusted by gateway can be used to invoke any of the API  protected with mutual ssl

### Describe your solution
After validating the certificate against the trust store , we should validate the certificate against the API currently being used. 

### How will you implement it
There will be list of certificate aliases maintained in the config agains the API. So that specific set of certificates are only allowed to access that API.

```toml
 # API 1
    [[mutualSSLConfig.api.certificates]]
      name = "Swagger Petstore"
      version = "1.0.5"
      aliasList = ["ballerina", "wso2apim310"]
    # API 2
    [[mutualSSLConfig.api.certificates]]
      name = "MyAPI"
      version = "2.0.0"
      aliasList = ["myAlias1", "myAlias2"]
```

---
#### Related Issues:
#988

### PR DESCRIPTION
**Pull request by tharmini**
### Purpose
Support mutual SSL in API level

### Issues
Fixes #988
Fixes #1269

If there is an LB between the client and the gateway, LB sets the client's certificate in a header. Then the Microgateway will check whether the mutual SSL handshake is passed. After that, Microgateway will check whether the request has the certificate header. 
    

1.  If it has the certificate header, then it validates that header certificate with the mutual SSL certificate configured for the API validated in the trustStore. 

2.  If it does not have the certificate header, the certificate used for the handshake is validated with the mutual SSL certificate configured for the API with the certificate alias.

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/LoadKeyStore.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package org.wso2.micro.gateway.core.mutualssl;
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/LoadKeyStore.java:1]: License header?

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/LoadKeyStore.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            trustStore.load(localTrustStoreStream, trustStorePassword.toCharArray());
+
+        } catch (NoSuchAlgorithmException | IOException | KeyStoreException | CertificateException e) {
+            String msg = "Error while loading the trustore";
+            log.error(msg, e);
+        }
+    }
+
+    /**
+     * Used to get the keystore path
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/LoadKeyStore.java:35]:
```suggestion
     * Used to get the keystore path.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithLoadBalancerHeader.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package org.wso2.micro.gateway.core.mutualssl;
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithLoadBalancerHeader.java:1]: License header?

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithLoadBalancerHeader.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            return getAliasFromTrustStore(x509Certificate, LoadKeyStore.trustStore);
+        } else {
+            return "";
+        }
+    }
+
+    /**
+     *  Used to get the certificate alias for a certificate exist in the trustore.
+     *
+     * @param certificate
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithLoadBalancerHeader.java:43]: Incomplete javadoc

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithoutLoadBalancerHeader.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package org.wso2.micro.gateway.core.mutualssl;
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithoutLoadBalancerHeader.java:1]: License header?

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/mutualSSL_handler.bal [hunk 6]
```diff
@@ -39,30 +56,107 @@ public type MutualSSLHandler object {
...
+                                    "API is not associated with the certificate");
+                                    setErrorMessageToInvocationContext(API_AUTH_INVALID_CREDENTIALS);
+                                    return false;
+                                } else {
+                                    printDebug(KEY_AUTHN_FILTER, "MutualSSL handshake status: PASSED");
+                                    doMTSLFilterRequest(req, invocationContext);
+                                    return true;
+                                }
+                            } else {
+                                handle|error aliasAFromHeaderCert = getAliasAFromHeaderCert(headerValue);
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/mutualSSL_handler.bal:90]: variable name aliasAFromHeaderCert , shouldn't it be aliasFromHeaderCert

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/mutualSSL_handler.bal [hunk 7]
```diff
@@ -39,30 +56,107 @@ public type MutualSSLHandler object {
...
+                            printDebug(KEY_AUTHN_FILTER, "Header has empty value sent by the payload");
+                            setErrorMessageToInvocationContext(API_AUTH_INVALID_CREDENTIALS);
+                            return false;
+                        }
+                    }
+                }
+                string? cert = req.mutualSslHandshake["base64EncodedCert"];
+                var cacheKey = cert.toString() + apiName + apiVersion;
+                var isExistCertCache = self.gatewayCache.retrieveFromMutualSslCertificateCache(cacheKey);
+                if (isExistCertCache is boolean) {
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/mutualSSL_handler.bal:123]: Why do we expect cert to be always present in the cache. If it is not in the cache shouldn't we go and check the cert in the trust store

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithoutLoadBalancerHeader.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import java.security.KeyStoreException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Base64;
+
+/**
+ * This class is for getting the certificate alias for a certificate to validate against per API.
+ */
+public class MutualsslWithoutLoadBalancerHeader {
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithoutLoadBalancerHeader.java:29]: Do we need two classes. Can't we use the same MutualsslWithLoadBalancerHeader. It has the same method. We can have a common class name and use it

File: components/micro-gateway-core/src/main/ballerina/src/gateway/utils/utils.bal [hunk 9]
```diff
@@ -905,8 +912,6 @@ function readMultipleJWTIssuers() {
     if (jwtIssuers is map<anydata>[] && jwtIssuers.length() > 0) {
         initiateJwtMap();
         printDebug(KEY_UTILS, "Found new multiple JWT issuer configs");
-        string trustStorePath = getConfigValue(LISTENER_CONF_INSTANCE_ID, TRUST_STORE_PATH, DEFAULT_TRUST_STORE_PATH);
-        string trustStorePassword = getConfigValue(LISTENER_CONF_INSTANCE_ID, TRUST_STORE_PASSWORD, DEFAULT_TRUST_STORE_PASSWORD);
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/utils/utils.bal:909]: Isn't these values are used below

Comment (tharmini) [reply components/micro-gateway-core/src/main/ballerina/src/gateway/utils/utils.bal:909]: I put this value above( calling for globally)

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithoutLoadBalancerHeader.java (hunk 8 above)
Comment (tharmini) [reply components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/MutualsslWithoutLoadBalancerHeader.java]: Added common class

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/CheckMutualSSLCertificate.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.cert.CertificateFactory;
+
+import javax.security.cert.CertificateEncodingException;
+import javax.security.cert.CertificateException;
+import javax.security.cert.X509Certificate;
+
+import static org.wso2.micro.gateway.core.Constants.BEGIN_CERTIFICATE_STRING;
+import static org.wso2.micro.gateway.core.Constants.END_CERTIFICATE_STRING;
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/CheckMutualSSLCertificate.java:33]: Lets not use static imports

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/CheckMutualSSLCertificate.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * This class is responsible for get the alias from the certificate  in mutual SSL handshake,
+ * when the header send by the load balancer.
+ */
+public class CheckMutualSSLCertificate {
+
+    public static String getAliasAFromHeaderCert(String base64EncodedCertificate) throws  KeyStoreException,
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/CheckMutualSSLCertificate.java:41]: lets rename the method getAliasFromHeaderCert

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/CheckMutualSSLCertificate.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+import javax.security.cert.X509Certificate;
+
+import static org.wso2.micro.gateway.core.Constants.BEGIN_CERTIFICATE_STRING;
+import static org.wso2.micro.gateway.core.Constants.END_CERTIFICATE_STRING;
+
+/**
+ * This class is responsible for get the alias from the certificate  in mutual SSL handshake,
+ * when the header send by the load balancer.
+ */
+public class CheckMutualSSLCertificate {
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/mutualssl/CheckMutualSSLCertificate.java:39]: Lets rename this class as CertificateUtils

**Pull request merged**